src/dice.py

```diff
@@ -0,0 +1,159 @@
+    d8 = 8
+    d10 = 10
+    d12 = 12
+    d20 = 20
+    d100 = 100
+
+
+class AdvantageSource(Enum):
+    terrain = "terrain"
+    stealth = "stealth"
+
+
+@dataclass(eq=True, frozen=True)
+class DieRollBonus:
+    source_description: str
+    amount: int
+
+
+@dataclass(eq=True, frozen=True)
+class DieRollMultiplier:
+    source_description: str
+    multiplier: float
+
+
+class RollResult:
+    def __init__(self, die: Die):
+        self._die_rolled: Die = die
+        self._natural_roll: List[int] = die.roll()
+        self._secondary_natural_roll: List[int] = die.roll()
+        self.__bonuses: Set[DieRollBonus] = set()
+        self.__multipliers: Set[DieRollMultiplier] = set()
+        self.tag: str | None = None
+
+    def add_bonus(self, bonus: DieRollBonus) -> None:
+        self.__bonuses.add(bonus)
+
+    def add_multiplier(self, multiplier: DieRollMultiplier) -> None:
+        self.__multipliers.add(multiplier)
+
+    def __get_final_multiplier(self):
+        if len(self.__multipliers) > 0:
+            return reduce(operator.mul, {mul.multiplier for mul in self.__multipliers})
+        else:
+            return 1
+
+    @property
+    def result(self) -> int | float:
+        unmultiplied_result = sum(self._natural_roll) + sum(bonus.amount for bonus in self.__bonuses)
+        return self.__get_final_multiplier() * unmultiplied_result
+
+    @property
+    def bonuses(self) -> Set[DieRollBonus]:
+        return self.__bonuses
+
+    def __repr__(self):
+        bonus_sum = sum(bonus.amount for bonus in self.__bonuses)
+        final_multiplier = self.__get_final_multiplier()
+        basic_repr = f"{self._die_rolled}+{bonus_sum}" if bonus_sum != 0 else self._die_rolled
+        return basic_repr if final_multiplier == 1 else f"{final_multiplier}x({basic_repr})"
+
+
+class StressDieRollResult(RollResult):
+    def __init__(self):
+        super().__init__(Die(1, DieType.d20))
+        self.__difficulty_class: int | None = None
+        self.__advantages: Set[AdvantageSource] = set()
+        self.__disadvantages: Set[AdvantageSource] = set()
+
+    def add_advantage(self, source: AdvantageSource) -> None:
+        self.__advantages.add(source)
+
+    def add_disadvantage(self, source: AdvantageSource) -> None:
+        self.__disadvantages.add(source)
+
+    @property
+    def has_advantage(self):
+        return len(self.__advantages) > len(self.__disadvantages)
+
+    @property
+    def has_disadvantage(self):
+        return len(self.__advantages) < len(self.__disadvantages)
+
+    def set_dc(self, dc: int) -> None:
+        self.__difficulty_class = dc
+
+    @property
+    def is_success(self) -> bool:
+        if self.__difficulty_class is None:
+            raise Exception("No DC was set.")
+        return self.result >= self.__difficulty_class
+
+    @property
+    def is_critical_success(self) -> bool:
+        if self.has_advantage:
+            return max(self._natural_roll[0], self._secondary_natural_roll[0]) == 20
```

Comment:
It might be more complex, but it is closer to natural language (which is why we love python)

```suggestion
            return any(roll == 20 for roll in [self._natural_roll[0], self._secondary_natural_roll[0]])
```

Reply:
Love it